File: website/js/search-engine.js
/**
 * Loopa Search Engine
 * Implements a prefix Trie and Levenshtein fuzzy string matching for client-side search indexing,
 * caching watchlist items, trending/popular items, and Gemini query corrections.
 */

class TrieNode {
    constructor() {
        this.children = {};
        this.titles = new Set(); // Stores full lowercase titles matching this prefix
    }
}

class Trie {
    constructor() {
        this.root = new TrieNode();
    }

    insert(word, title) {
        let node = this.root;
        for (const char of word) {
            if (!node.children[char]) {
                node.children[char] = new TrieNode();
            }
            node = node.children[char];
            node.titles.add(title);
        }
    }

    search(prefix) {
        let node = this.root;
        for (const char of prefix) {
            if (!node.children[char]) return [];
            node = node.children[char];
        }
        return Array.from(node.titles);
    }
}

const LoopaSearchEngine = {
    trie: new Trie(),
    mediaMap: new Map(), // lowercase title -> original MediaItem
    queryCorrectionCache: new Map(), // raw query -> corrected query
    stopwords: new Set(['the', 'and', 'for', 'with', 'a', 'an', 'of', 'in', 'to', 'is', 'on', 'at']),

    /**
     * Clear the search engine index and media map.
     */
    clearIndex() {
        this.trie = new Trie();
        this.mediaMap.clear();
    },

    /**
     * Index a single MediaItem.
     * @param {Object} item - Normalized MediaItem
     */
    indexMediaItem(item) {
        if (!item || !item.title) return;
        const fullTitle = item.title.toLowerCase().trim();
        
        // Map lowercase title to original item (avoid overwriting higher-scored items if possible)
        const existing = this.mediaMap.get(fullTitle);
        if (!existing || (item.score && (!existing.score || item.score > existing.score))) {
            this.mediaMap.set(fullTitle, item);
        }

        // 1. Index full title as a single phrase
        const cleanTitle = fullTitle.replace(/[^a-z0-9\s]/g, '').replace(/\s+/g, ' ').trim();
        if (cleanTitle) {
            this.trie.insert(cleanTitle, fullTitle);
        }

        // 2. Index individual words to allow sub-word matching (e.g. "japan" matching "Doraemon Japan")
        const words = cleanTitle.split(' ');
        words.forEach(word => {
            if (word.length >= 3 && !this.stopwords.has(word)) {
                this.trie.insert(word, fullTitle);
            }
        });
    },

    /**
     * Index a list of MediaItems.
     * @param {Array} items - Array of Normalized MediaItems
     */
    indexMediaItems(items) {
        if (!Array.isArray(items)) return;
        items.forEach(item => this.indexMediaItem(item));
        console.log(`[SearchEngine] Indexed ${items.length} items. Total unique targets: ${this.mediaMap.size}`);
    },

    /**
     * Calculate Levenshtein Distance between two strings.
     */
    levenshteinDistance(str1, str2) {
        const track = Array(str2.length + 1).fill(null).map(() => Array(str1.length + 1).fill(null));
        for (let i = 0; i <= str1.length; i += 1) track[0][i] = i;
        for (let j = 0; j <= str2.length; j += 1) track[j][0] = j;
        for (let j = 1; j <= str2.length; j += 1) {
            for (let i = 1; i <= str1.length; i += 1) {
                const indicator = str1[i - 1] === str2[j - 1] ? 0 : 1;
                track[j][i] = Math.min(
                    track[j][i - 1] + 1, // deletion
                    track[j - 1][i] + 1, // insertion
                    track[j - 1][i - 1] + indicator // substitution
                );
            }
        }
        return track[str2.length][str1.length];
    },

    /**
     * Query autocomplete suggestions.
     * Combines Trie prefix matches, sub-word prefix matches, and falls back to Levenshtein fuzzy match.
     * @param {string} rawQuery
     * @param {number} limit
     * @returns {Array<Object>} List of matched MediaItems
     */
    getSuggestions(rawQuery, limit = 5) {
        const query = rawQuery.toLowerCase().trim().replace(/[^a-z0-9\s]/g, '').replace(/\s+/g, ' ');
        if (!query) return [];

        // 1. Attempt prefix search
        const matchedTitles = this.trie.search(query);
        let results = [];
        const seen = new Set();

        matchedTitles.forEach(t => {
            if (seen.size >= limit) return;
            const item = this.mediaMap.get(t);
            if (item && !seen.has(item.id + '_' + item.mediaType)) {
                results.push(item);
                seen.add(item.id + '_' + item.mediaType);
            }
        });

        // 2. If results are less than limit, try individual word prefix matches
        if (results.length < limit) {
            const queryWords = query.split(' ');
            if (queryWords.length > 1) {
                // Find items matching all query words as prefixes
                for (const [title, item] of this.mediaMap.entries()) {
                    if (seen.size >= limit) break;
                    if (seen.has(item.id + '_' + item.mediaType)) continue;

                    const titleWords = title.split(/\s+/);
                    const allWordsMatch = queryWords.every(qWord => 
                        titleWords.some(tWord => tWord.startsWith(qWord))
                    );

                    if (allWordsMatch) {
                        results.push(item);
                        seen.add(item.id + '_' + item.mediaType);
                    }
                }
            }
        }

        // 3. Fallback: Fuzzy matching using Levenshtein distance on words
        if (results.length < limit) {
            const fuzzyCandidates = [];
            const maxErrors = Math.min(2, Math.floor(query.length / 2));

            for (const [title, item] of this.mediaMap.entries()) {
                if (seen.has(item.id + '_' + item.mediaType)) continue;

                // Simple check: is Levenshtein distance of the entire string close?
                const dist = this.levenshteinDistance(query, title);
                if (dist <= maxErrors) {
                    fuzzyCandidates.push({ item, score: dist, type: 'full' });
                    continue;
                }

                // Check Levenshtein distance of individual words
                const titleWords = title.replace(/[^a-z0-9\s]/g, '').split(/\s+/);
                const queryWords = query.split(' ');
                let matchFound = false;

                for (const qWord of queryWords) {
                    if (qWord.length < 3) continue;
                    for (const tWord of titleWords) {
                        if (tWord.length < 3) continue;
                        const wordDist = this.levenshteinDistance(qWord, tWord);
                        const wordMaxErrors = Math.min(1, Math.floor(qWord.length / 3));
                        if (wordDist <= wordMaxErrors) {
                            fuzzyCandidates.push({ item, score: wordDist, type: 'word' });
                            matchFound = true;
                            break;
                        }
                    }
                    if (matchFound) break;
                }
            }

            // Sort fuzzy matches: lowest distance first, full phrase match preferred over word matches
            fuzzyCandidates.sort((a, b) => {
                if (a.type !== b.type) return a.type === 'full' ? -1 : 1;
                return a.score - b.score;
            });

            fuzzyCandidates.forEach(cand => {
                if (seen.size >= limit) return;
                results.push(cand.item);
                seen.add(cand.item.id + '_' + cand.item.mediaType);
            });
        }

        return results;
    },

    /**
     * Cache a corrected query mapping.
     */
    cacheCorrection(query, corrected) {
        if (!query || !corrected) return;
        this.queryCorrectionCache.set(query.toLowerCase().trim(), corrected.trim());
    },

    /**
     * Retrieve cached correction.
     */
    getCachedCorrection(query) {
        if (!query) return null;
        return this.queryCorrectionCache.get(query.toLowerCase().trim()) || null;
    }
};

window.LoopaSearchEngine = LoopaSearchEngine;

File: app/src/main/java/com/loopa/search/SearchEngine.kt
package com.loopa.search

import com.loopa.model.TmdbMovie
import com.loopa.db.MediaItemEntity
import java.util.Locale
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.CopyOnWriteArrayList

class TrieNode {
    val children = ConcurrentHashMap<Char, TrieNode>()
    val titles = ConcurrentHashMap.newKeySet<String>() // Stores full lowercase titles matching this prefix
}

class Trie {
    private val root = TrieNode()

    fun insert(word: String, title: String) {
        var node = root
        for (char in word) {
            node = node.children.computeIfAbsent(char) { TrieNode() }
            node.titles.add(title)
        }
    }

    fun search(prefix: String): List<String> {
        var node = root
        for (char in prefix) {
            node = node.children[char] ?: return emptyList()
        }
        return node.titles.toList()
    }
}

object SearchEngine {
    private val trie = Trie()
    private val mediaMap = ConcurrentHashMap<String, TmdbMovie>() // lowercase title -> original TmdbMovie
    private val queryCorrectionCache = ConcurrentHashMap<String, String>() // raw query -> corrected query
    private val stopwords = setOf("the", "and", "for", "with", "a", "an", "of", "in", "to", "is", "on", "at")

    fun clearIndex() {
        // Since we can't easily recreate a private val reference directly or clear it, we recreate node links
        // We can just wipe out the Trie by creating a new Trie instance or clearing the maps
        // Let's make Trie class mutable or instantiate a new one. We can do:
        synchronized(this) {
            // Re-instantiate Trie is fine if we make it a var. Let's make it a var!
            _trie = Trie()
            mediaMap.clear()
        }
    }

    private var _trie = Trie()

    fun indexMediaItem(item: TmdbMovie) {
        val title = item.title ?: item.name ?: return
        val fullTitle = title.lowercase(Locale.getDefault()).trim()
        if (fullTitle.isEmpty()) return

        val existing = mediaMap[fullTitle]
        if (existing == null || (item.voteAverage ?: 0.0) > (existing.voteAverage ?: 0.0)) {
            mediaMap[fullTitle] = item
        }

        // 1. Index full title as a single phrase
        val cleanTitle = fullTitle.replace(Regex("[^a-z0-9\\s]"), "").replace(Regex("\\s+"), " ").trim()
        if (cleanTitle.isNotEmpty()) {
            _trie.insert(cleanTitle, fullTitle)
        }

        // 2. Index individual words
        val words = cleanTitle.split(" ")
        for (word in words) {
            if (word.length >= 3 && !stopwords.contains(word)) {
                _trie.insert(word, fullTitle)
            }
        }
    }

    fun indexMediaItems(items: List<TmdbMovie>) {
        for (item in items) {
            indexMediaItem(item)
        }
    }

    fun indexMediaItemEntity(entity: MediaItemEntity) {
        val movie = TmdbMovie(
            id = entity.id,
            title = entity.title,
            name = null,
            overview = entity.personalNotes,
            posterPath = entity.imageUrl?.replace("https://image.tmdb.org/t/p/w500", "")?.replace("https://loopa-tmdb-proxy.sujalsanjay-chhajed2023.workers.dev/t/p/w500", ""),
            backdropPath = null,
            voteAverage = entity.score,
            releaseDate = entity.date,
            firstAirDate = null,
            mediaType = entity.mediaType,
            popularity = 0.0,
            genreIds = null
        )
        indexMediaItem(movie)
    }

    fun indexMediaItemEntities(entities: List<MediaItemEntity>) {
        for (entity in entities) {
            indexMediaItemEntity(entity)
        }
    }

    private fun levenshteinDistance(str1: String, str2: String): Int {
        val dp = Array(str2.length + 1) { IntArray(str1.length + 1) }
        for (i in 0..str1.length) dp[0][i] = i
        for (j in 0..str2.length) dp[j][0] = j
        for (j in 1..str2.length) {
            for (i in 1..str1.length) {
                val indicator = if (str1[i - 1] == str2[j - 1]) 0 else 1
                dp[j][i] = minOf(
                    dp[j][i - 1] + 1, // deletion
                    dp[j - 1][i] + 1, // insertion
                    dp[j - 1][i - 1] + indicator // substitution
                )
            }
        }
        return dp[str2.length][str1.length]
    }

    fun getSuggestions(rawQuery: String, limit: Int = 5): List<TmdbMovie> {
        val query = rawQuery.lowercase(Locale.getDefault()).trim().replace(Regex("[^a-z0-9\\s]"), "").replace(Regex("\\s+"), " ")
        if (query.isEmpty()) return emptyList()

        val matchedTitles = _trie.search(query)
        val results = CopyOnWriteArrayList<TmdbMovie>()
        val seen = ConcurrentHashMap.newKeySet<String>()

        for (t in matchedTitles) {
            if (seen.size >= limit) break
            val item = mediaMap[t]
            if (item != null) {
                val key = "${item.id}_${item.mediaType ?: "movie"}"
                if (seen.add(key)) {
                    results.add(item)
                }
            }
        }

        // 2. If results < limit, try individual word prefix matches
        if (results.size < limit) {
            val queryWords = query.split(" ")
            if (queryWords.size > 1) {
                for ((title, item) in mediaMap.entries) {
                    if (seen.size >= limit) break
                    val key = "${item.id}_${item.mediaType ?: "movie"}"
                    if (seen.contains(key)) continue

                    val titleWords = title.split(Regex("\\s+"))
                    val allWordsMatch = queryWords.all { qWord ->
                        titleWords.any { tWord -> tWord.startsWith(qWord) }
                    }

                    if (allWordsMatch) {
                        if (seen.add(key)) {
                            results.add(item)
                        }
                    }
                }
            }
        }

        // 3. Levenshtein fuzzy match
        if (results.size < limit) {
            val fuzzyCandidates = CopyOnWriteArrayList<FuzzyCandidate>()
            val maxErrors = minOf(2, query.length / 2)

            for ((title, item) in mediaMap.entries) {
                val key = "${item.id}_${item.mediaType ?: "movie"}"
                if (seen.contains(key)) continue

                val dist = levenshteinDistance(query, title)
                if (dist <= maxErrors) {
                    fuzzyCandidates.add(FuzzyCandidate(item, dist, "full"))
                    continue
                }

                val titleWords = title.replace(Regex("[^a-z0-9\\s]"), "").split(Regex("\\s+"))
                val queryWords = query.split(" ")
                var matchFound = false

                for (qWord in queryWords) {
                    if (qWord.length < 3) continue
                    for (tWord in titleWords) {
                        if (tWord.length < 3) continue
                        val wordDist = levenshteinDistance(qWord, tWord)
                        val wordMaxErrors = minOf(1, qWord.length / 3)
                        if (wordDist <= wordMaxErrors) {
                            fuzzyCandidates.add(FuzzyCandidate(item, wordDist, "word"))
                            matchFound = true
                            break
                        }
                    }
                    if (matchFound) break
                }
            }

            // Sort candidates
            val sorted = fuzzyCandidates.sortedWith(compareBy<FuzzyCandidate> { it.type != "full" }.thenBy { it.score })
            for (cand in sorted) {
                if (seen.size >= limit) break
                val key = "${cand.item.id}_${cand.item.mediaType ?: "movie"}"
                if (seen.add(key)) {
                    results.add(cand.item)
                }
            }
        }

        return results.toList()
    }

    fun cacheCorrection(query: String, corrected: String) {
        val q = query.lowercase(Locale.getDefault()).trim()
        val c = corrected.trim()
        if (q.isNotEmpty() && c.isNotEmpty()) {
            queryCorrectionCache[q] = c
        }
    }

    fun getCachedCorrection(query: String): String? {
        val q = query.lowercase(Locale.getDefault()).trim()
        return queryCorrectionCache[q]
    }

    private data class FuzzyCandidate(val item: TmdbMovie, val score: Int, val type: String)
}
